### PR TITLE
[USM] protocol classification: add RabbitMQ classification 

### PR DESCRIPTION
**Pull request by amitslavin**
### What does this PR do?
Adds RabbitMQ to USM protocol classification.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/network/ebpf/c/protocols/protocol-classification-defs.h [hunk 1]
```diff
@@ -25,6 +42,7 @@ typedef enum {
     PROTOCOL_HTTP,
     PROTOCOL_HTTP2,
     PROTOCOL_TLS,
+    PROTOCOL_AMQP,
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-defs.h:45]: that's wrong
you should open a PR in the [agent-payload repo](https://github.com/DataDog/agent-payload/blob/master/proto/process/agent.proto) to add the protocol to the list.
the value here should match the value in the protobuf.

File: pkg/network/ebpf/c/protocols/protocol-classification-defs.h [hunk 2]
```diff
@@ -15,6 +15,23 @@
...
+#define AMQP_METHOD_CONNECTION_START_OK 11
+
+// RabbitMQ supported methods types.
+#define AMQP_METHOD_CONSUME 20
+#define AMQP_METHOD_PUBLISH 40
+#define AMQP_METHOD_DELIVER 60
+#define AMQP_FRAME_METHOD_TYPE 1
+
+#define AMQP_MIN_FRAME_LENGTH 8
+#define AMQP_MIN_PROTOOL_HEADER_SIZE 4
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-defs.h:33]: please add a reference to the protocol RFC

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 3]
```diff
@@ -36,6 +36,18 @@ static __always_inline bool is_http2_preface(const char* buf, __u32 buf_size) {
     return match;
 }
 
+// The method checks if the given buffer starts with the HTTP2 marker as defined in https://datatracker.ietf.org/doc/html/rfc7540.
+// We check that the given buffer is not empty and its size is at least 24 bytes.
+static __always_inline bool is_amqp_protocol_header(const char* buf, __u32 buf_size) {
+    CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, AMQP_MIN_PROTOOL_HEADER_SIZE)
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:42]: should be 8, according to the RFC, the protocol starts with 8 octets

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 4]
```diff
@@ -36,6 +36,18 @@ static __always_inline bool is_http2_preface(const char* buf, __u32 buf_size) {
     return match;
 }
 
+// The method checks if the given buffer starts with the HTTP2 marker as defined in https://datatracker.ietf.org/doc/html/rfc7540.
+// We check that the given buffer is not empty and its size is at least 24 bytes.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:40]: modify the documentation

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 5]
```diff
@@ -57,6 +69,51 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
     return is_http2_preface(buf, buf_size) || is_http2_server_settings(buf, buf_size);
 }
 
+// The method checks if the given buffer starts is amqp message.
+// Ref: https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf
+static __always_inline bool is_amqp(const char* buf, __u32 buf_size) {
+    if (buf_size < AMQP_MIN_FRAME_LENGTH) {
+        return false;
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:77]:
```suggestion
    CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, AMQP_MIN_FRAME_LENGTH)
```

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 6]
```diff
@@ -57,6 +69,51 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+
+    uint8_t frame_type = buf[0];
+    // Check only for method frame type.
+    if (frame_type != AMQP_FRAME_METHOD_TYPE) {
+        return false;
+    }
+
+    __u16 class_id = buf[7] << 8 | buf[8];
+    __u16 method_id = buf[9] << 8 | buf[10];
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:92]: explain what are we doing here, and why

Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:92]: furthermore, you checked the buffer size if not less than `AMQP_MIN_FRAME_LENGTH` (8), but here you are accessing the positions `8`, `9` & `10` which exceed from the length you verified

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 7]
```diff
@@ -57,6 +69,51 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+    // ConnectionStart, ConnectionStartOk, BasicPublish, BasicDeliver, BasicConsume are the most likely methods to
+    // consider for the classification.
+    if (class_id == AMQP_CONNECTION_CLASS && method_id == AMQP_METHOD_CONNECTION_START) {
+        return true;
+    }
+    if (class_id == AMQP_CONNECTION_CLASS && method_id == AMQP_METHOD_CONNECTION_START_OK) {
+        return true;
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:100]:
```suggestion
    if (class_id == AMQP_CONNECTION_CLASS) {
        return  method_id == AMQP_METHOD_CONNECTION_START || method_id == AMQP_METHOD_CONNECTION_START_OK;
    }
```

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 8]
```diff
@@ -57,6 +69,51 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+
+    if (class_id == AMQP_BASIC_CLASS && method_id == AMQP_METHOD_DELIVER) {
+        return true;
+    }
+
+    if (class_id == AMQP_BASIC_CLASS && method_id == AMQP_METHOD_CONSUME) {
+        return true;
+    }
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:112]:
```suggestion
    if (class_id == AMQP_BASIC_CLASS) {
        return method_id == AMQP_METHOD_PUBLISH || method_id == AMQP_METHOD_DELIVER || method_id == AMQP_METHOD_CONSUME;
    }
```

File: pkg/network/event_common.go [hunk 9]
```diff
@@ -26,6 +26,7 @@ const (
 	ProtocolHTTP
 	ProtocolHTTP2
 	ProtocolTLS
+	ProtocolAMQP
```
Comment (guyarb) [inline pkg/network/event_common.go:29]: again, should match the protobuf value

File: pkg/network/tracer/testutil/amqp/server.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+func RunAmqp(t *testing.T, serverAddr string) {
+	t.Helper()
+
+	env := []string{
+		"AMQP_ADDR=" + serverAddr,
+	}
+	dir, _ := testutil.CurDir()
+	cmd := exec.Command("docker-compose", "-f", dir+"/testdata/docker-compose.yml", "up", "-d")
+	cmd.Stderr = os.Stdout
+	cmd.Stdout = os.Stdout
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/server.go:20]: remove, that's for debugging purposes

File: pkg/network/tracer/testutil/amqp/server.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package amqp
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
+	"github.com/stretchr/testify/require"
+	"os"
+	"os/exec"
+	"testing"
+)
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/server.go:9]:
```suggestion
import (
	"os"
	"os/exec"
	"testing"
	
	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
	"github.com/stretchr/testify/require"
)
```

File: pkg/network/tracer/testutil/amqp/server.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
+package amqp
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/server.go:1]: missing copyrights

Comment (guyarb) [reply pkg/network/tracer/testutil/amqp/server.go:1]: please locate that directory under `/pkg/network/protocols`

File: pkg/network/tracer/testutil/amqp/server.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
+	"github.com/stretchr/testify/require"
+	"os"
+	"os/exec"
+	"testing"
+)
+
+func RunAmqp(t *testing.T, serverAddr string) {
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/server.go:11]:
```suggestion
func RunAmqpServer(t *testing.T, serverAddr string) {
```

Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/server.go:11]: allow to decide the port at well

File: pkg/network/tracer/testutil/amqp/sender.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
+package amqp
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/sender.go:1]: missing copyrights

File: pkg/network/tracer/testutil/amqp/consumer.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
+package amqp
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/consumer.go:1]: missing copyrights

File: pkg/network/tracer/testutil/amqp/consumer.go [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+)
+
+func failOnError(err error, msg string) {
+	if err != nil {
+		log.Fatalf("%s: %s", msg, err)
+	}
+}
+
+func ConsumeAmqp() {
+	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/consumer.go:17]: the address and port should be configurable

File: pkg/network/tracer/tracer_classification_test.go [hunk 17]
```diff
@@ -175,6 +178,46 @@ func testProtocolClassification(t *testing.T, cfg *config.Config, clientHost, ta
...
+
+				return false, ""
+			},
+		},
+		{
+			name: "amqp consumer",
+			want: network.ProtocolAMQP,
+			clientRun: func(t *testing.T, serverAddr string) {
+				time.Sleep(5 * time.Second)
+				amqp.ConsumeAmqp()
```
Comment (guyarb) [inline pkg/network/tracer/tracer_classification_test.go:206]: you don't use serverAddr

File: pkg/network/tracer/testutil/amqp/sender.go [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+// Here we set the way error messages are displayed in the terminal.
+func logError(err error, msg string) {
+	if err != nil {
+		log.Fatalf("%s: %s", msg, err)
+	}
+}
+
+func Send() {
+	// Here we connect to RabbitMQ or send a message if there are any errors connecting.
+	conn, err := amqp.Dial("amqp://guest:guest@localhost:5672/")
```
Comment (guyarb) [inline pkg/network/tracer/testutil/amqp/sender.go:18]: the address and port should be configurable

File: pkg/network/tracer/tracer_classification_test.go [hunk 19]
```diff
@@ -12,15 +12,17 @@ import (
 	"bufio"
 	"context"
 	"fmt"
+	"github.com/DataDog/datadog-agent/pkg/network/tracer/testutil/amqp"
+	"github.com/DataDog/datadog-agent/pkg/network/tracer/testutil/grpc"
```
Comment (guyarb) [inline pkg/network/tracer/tracer_classification_test.go:16]: organize the imports

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/ebpf/c/protocols/protocol-classification-defs.h (hunk 1 above)
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/protocol-classification-defs.h:45]: done: https://github.com/DataDog/agent-payload/pull/210

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h (hunk 6 above)
Comment (amitslavin) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:92]: the min frame length is indeed 8 which may include header and a payload, if we got to the part which we extract the class id and the method id it is definitely part of the "method frame" which should be bigger then 8 for sure and we should be able to extract those fields

File: go.mod [hunk 20]
```diff
@@ -53,7 +53,7 @@ require (
 	github.com/DataDog/datadog-agent/pkg/util/log v0.42.0-rc.2
 	github.com/DataDog/datadog-agent/pkg/util/scrubber v0.42.0-rc.2
 	github.com/DataDog/datadog-go/v5 v5.1.1
-	github.com/DataDog/datadog-operator v0.7.1-0.20220602134901-4f6af09bf54f
+	github.com/DataDog/datadog-operator v0.8.3
```
Comment (guyarb) [inline go.mod:56]: I'd revert that change

File: pkg/network/ebpf/c/protocols/protocol-classification-defs.h (hunk 1 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-defs.h:45]: still, the value here will be `5` while it should be `8`

```suggestion
    PROTOCOL_AMQP = 8,
```

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 21]
```diff
@@ -36,6 +36,18 @@ static __always_inline bool is_http2_preface(const char* buf, __u32 buf_size) {
...
+static __always_inline bool is_amqp_protocol_header(const char* buf, __u32 buf_size) {
+    CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, AMQP_MIN_FRAME_LENGTH)
+
+#define AMQP_PREFACE "AMQP"
+
+    bool match = !bpf_memcmp(buf, AMQP_PREFACE, sizeof(AMQP_PREFACE)-1);
+
+    return match;
+}
+
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:50]: nit: move the amqp classification methods to be next to each other

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 22]
```diff
@@ -57,6 +69,41 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
     return is_http2_preface(buf, buf_size) || is_http2_server_settings(buf, buf_size);
 }
 
+// The method checks if the given buffer starts is amqp message.
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:72]:
```suggestion
// The method checks if the given buffer is an AMQP message.
```

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 23]
```diff
@@ -57,6 +69,41 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+// Ref: https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf
+static __always_inline bool is_amqp(const char* buf, __u32 buf_size) {
+    CHECK_PRELIMINARY_BUFFER_CONDITIONS(buf, buf_size, AMQP_MIN_FRAME_LENGTH)
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:75]: redundant, as the same line exists in `is_amqp_protocol_header`

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 24]
```diff
@@ -57,6 +69,41 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+
+    // New connection should start with protocol header of AMQP.
+    // Ref https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf.
+    if (is_amqp_protocol_header(buf, buf_size)) {
+        return true;
+    }
+
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:82]: Once we know it is not a start of AMQP, we should check the size fits to our needs.
since you are reading up to the position `10` in the array, you must check the buffer size is at least `11`
So add
```
#define AMQP_MIN_PAYLOAD_LENGTH 11
```

```suggestion
    if (is_amqp_protocol_header(buf, buf_size)) {
        return true;
    }

    if (buf_size < AMQP_MIN_PAYLOAD_LENGTH) {
        return false;
    }
```

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h (hunk 6 above)
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:92]: I meant that you are right, in the protocol the minimum message can be 8 bytes, but here, you are assuming that the positions `8`, `9` & `10` are valid to access in the buffer, so you are expecting to have payloads with length of at least 11 characters.

File: pkg/network/protocols/amqp/consumer.go [hunk 25]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/streadway/amqp"
+)
+
+func failOnError(err error, msg string) {
+	if err != nil {
+		log.Fatalf("%s: %s", msg, err)
+	}
+}
+
+func ConsumeAmqp(serverAddr string, serverPort string) {
```
Comment (guyarb) [inline pkg/network/protocols/amqp/consumer.go:22]: nit:
```suggestion
func ConsumeAmqp(serverAddr, serverPort string) {
```

File: pkg/network/protocols/amqp/sender.go [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package amqp
+
+import (
+	"fmt"
+	"github.com/streadway/amqp"
+	"log"
+)
```
Comment (guyarb) [inline pkg/network/protocols/amqp/sender.go:12]: nit:

```suggestion
import (
	"fmt"
	"log"

	"github.com/streadway/amqp"	
)
```

File: pkg/network/protocols/amqp/server.go [hunk 27]
```diff
@@ -0,0 +1,38 @@
...
+
+package amqp
+
+import (
+	"os"
+	"os/exec"
+	"testing"
+
+	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
+	"github.com/stretchr/testify/require"
```
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:14]: nit:

```suggestion
	"github.com/stretchr/testify/require"

	"github.com/DataDog/datadog-agent/pkg/network/protocols/http/testutil"
```

File: pkg/network/protocols/amqp/server.go [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+)
+
+func RunAmqpServer(t *testing.T, serverAddr string, serverPort string) {
```
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:17]: nit:

```suggestion
func RunAmqpServer(t *testing.T, serverAddr, serverPort string) {
```

File: pkg/network/protocols/amqp/server.go [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+	t.Helper()
+
+	env := []string{
+		"AMQP_ADDR=" + serverAddr,
+		"AMQP_PORT=" + serverPort,
+	}
+	dir, _ := testutil.CurDir()
+	cmd := exec.Command("docker-compose", "-f", dir+"/testdata/docker-compose.yml", "up", "-d")
+	cmd.Stderr = os.Stdout
+	cmd.Stdout = os.Stdout
```
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:27]: remove

File: pkg/network/protocols/amqp/server.go [hunk 30]
```diff
@@ -0,0 +1,38 @@
...
+	cmd.Env = append(cmd.Env, env...)
+	require.NoErrorf(t, cmd.Run(), "could not start amqp with docker-compose")
+
+	t.Cleanup(func() {
+		c := exec.Command("docker-compose", "-f", dir+"/testdata/docker-compose.yml", "down", "--remove-orphans")
+		c.Stdout = os.Stdout
+		c.Stderr = os.Stderr
```
Comment (guyarb) [inline pkg/network/protocols/amqp/server.go:34]: remove

File: pkg/network/protocols/amqp/testdata/docker-compose.yml [hunk 31]
```diff
@@ -0,0 +1,7 @@
+version: '3'
```
Comment (guyarb) [inline pkg/network/protocols/amqp/testdata/docker-compose.yml:1]: TODO: Add the docker compose to system-probe.py once https://github.com/DataDog/datadog-agent/pull/14589 is merged

File: pkg/network/tracer/tracer_classification_test.go [hunk 32]
```diff
@@ -175,6 +178,51 @@ func testProtocolClassification(t *testing.T, cfg *config.Config, clientHost, ta
 			},
 			want: network.ProtocolHTTP2,
 		},
+		{
+			name: "amqp sender",
+			want: network.ProtocolAMQP,
+			clientRun: func(t *testing.T, serverAddr string) {
+				time.Sleep(5 * time.Second)
+				addr, _, _ := net.SplitHostPort(serverAddr)
+				port := "5672"
```
Comment (guyarb) [inline pkg/network/tracer/tracer_classification_test.go:187]: TODO: fix that

**Review by guyarb — CHANGES_REQUESTED** [review]
Looks great!

File: pkg/network/protocols/amqp/server.go (hunk 27 above)
Comment (amitslavin) [reply pkg/network/protocols/amqp/server.go:14]: when I change it to the suggest order it marks it as warning, then I do the suggested change it it returns to the shape I committed.

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 33]
```diff
@@ -69,17 +57,64 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+        && bpf_memcmp(buf, POST, sizeof(POST)-1)
+        && bpf_memcmp(buf, PUT, sizeof(PUT)-1)
+        && bpf_memcmp(buf, DELETE, sizeof(DELETE)-1)
+        && bpf_memcmp(buf, HEAD, sizeof(HEAD)-1)
+        && bpf_memcmp(buf, OPTIONS1, sizeof(OPTIONS1)-1)
+        && bpf_memcmp(buf, OPTIONS2, sizeof(OPTIONS2)-1)
+        && bpf_memcmp(buf, PATCH, sizeof(PATCH)-1));
+
+    return http;
+}
```
Comment (guyarb) [inline pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:90]: Please revert the change

**Review by guyarb — CHANGES_REQUESTED** [review]

File: pkg/network/protocols/amqp/server.go (hunk 30 above)
Comment (guyarb) [reply pkg/network/protocols/amqp/server.go:32]: still exist

File: pkg/network/event_usm_common.go [hunk 34]
```diff
@@ -14,4 +14,5 @@ const (
 	ProtocolHTTP2
 	ProtocolTLS
 	MaxProtocols
+	ProtocolAMQP = 8
```
Comment (guyarb) [inline pkg/network/event_usm_common.go:17]:
```suggestion
    ProtocolAMQP = 8
	MaxProtocols = 9
```

**Review by guyarb — CHANGES_REQUESTED** [review]

**Review by guyarb — APPROVED** [review]

**Review by usamasaqib — APPROVED** [review]

File: pkg/network/event_usm_common.go [hunk 35]
```diff
@@ -13,5 +13,6 @@ const (
 	ProtocolHTTP
 	ProtocolHTTP2
 	ProtocolTLS
-	MaxProtocols
+	ProtocolAMQP = 8
```
Comment (nplanel) [inline pkg/network/event_usm_common.go:16]: we adding a gap between TLS and AMQP id ?

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 36]
```diff
@@ -57,6 +57,56 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+// Ref: https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf
+static __always_inline bool is_amqp(const char* buf, __u32 buf_size) {
+    // New connection should start with protocol header of AMQP.
+    // Ref https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf.
+    if (is_amqp_protocol_header(buf, buf_size)) {
+        return true;
+    }
+
+    // Validate that we will be able to get from the buffer the class and method ids.
+    if (buf_size < AMQP_MIN_PAYLOAD_LENGTH) {
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:82]: no, `is_amqp` calls for `is_amqp_protocol_header` at the begging
in that function (and according to the protocol) we might get 8 bytes message that starts with `AMQP` string.
in `is_amqp_protocol_header` we check that the size is at least 8 bytes, and if not, we harden the check to make sure it fits the other mechanism we look for

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 37]
```diff
@@ -99,6 +149,8 @@ static __always_inline void classify_protocol(protocol_t *protocol, const char *
         *protocol = PROTOCOL_HTTP;
     } else if (is_http2(buf, size)) {
         *protocol = PROTOCOL_HTTP2;
+    } else if (is_amqp(buf, size)) {
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:152]: that's incorrect
1. the methods will be called if we didn't classify the traffic, meaning that once we identified the protocol for the connection, we will never try to reclassify it
2. the same connection can have multiple requests / responses. so you cannot assume that you read the first packet of the response

File: pkg/network/event_usm_common.go (hunk 35 above)
Comment (guyarb) [reply pkg/network/event_usm_common.go:16]: the list here represents the same list in `agent-payload` and there we have other protocol in that gap.
each PR adds the protocol it represents

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h [hunk 38]
```diff
@@ -57,6 +57,56 @@ static __always_inline bool is_http2(const char* buf, __u32 buf_size) {
...
+
+    uint8_t frame_type = buf[0];
+    // Check only for method frame type.
+    if (frame_type != AMQP_FRAME_METHOD_TYPE) {
+        return false;
+    }
+
+    // We extract the class id and method id by big indian from the buffer.
+    // Ref https://www.rabbitmq.com/resources/specs/amqp0-9-1.pdf.
+    __u16 class_id = buf[7] << 8 | buf[8];
```
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:94]: the scope of the PR is for the first version of the protocol
we do plan to support version 0.10 and 1.0 in a different PR

[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:94]: ok thanks

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h (hunk 37 above)
[the remark this answers is not in the document]
Comment (nplanel) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:152]: I talking about the path where no protocol as been classified for x packets, we will still continue to try to classified and possibly miss classified flow.  (ex: try to classified encrypted, mean unknown)

File: pkg/network/event_usm_common.go (hunk 35 above)
Comment (nplanel) [reply pkg/network/event_usm_common.go:16]: why not using type aliasing here with agent payload ProtocolType
type ProtocolType agentpb.ProtocolType
const (
    ProtocolHTTP2 = ProtocolType_protocolHTTP2
...
)

Comment (guyarb) [reply pkg/network/event_usm_common.go:16]: won't work on windows

File: pkg/network/ebpf/c/protocols/protocol-classification-helpers.h (hunk 37 above)
[the remark this answers is not in the document]
Comment (guyarb) [reply pkg/network/ebpf/c/protocols/protocol-classification-helpers.h:152]: > I talking about the path where no protocol as been classified for x packets
we've talked about that in the past, there are 2 main issues with the approach:
1. we need to hold state for all connections even if we didn't classify them
2. we are not covering all options in the classification, meaning that we can choose couple of operations and classify by them. for example in kafka, we classify by fetch/produce messages, but there are 64 other possible messages in the protocol. we can miss connections in your suggestion if we see other messages (For example metadata request, or list topics)

> we will still continue to try to classified and possibly miss classified flow

I didn't understand your intention here, can explain?


we've have been talking that in the past, but
